### PR TITLE
Add node status check for OpenBMC rpower status

### PR DESCRIPTION
**Pull request by xuweibj**
After ``rpower off``, the status of node is ``powering-off``. 
* If we get status that chassis status "Off", Host status "Off" and Host Transition status "On", check the status of node(getNodesAttribs). If it's "powering-off", the current status is "OFF". If not, the current status is "powering-on".
* If we get status that chassis staus "On", Host status "Running" and Host Transition status "On", but the status of node is "powering-off", the current status is "powering-off".

Run ``rpower off``
```
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn11 stat
mid05tor12cn11: DEBUG State CurrentBMCState=xyz.openbmc_project.State.BMC.BMCState.Ready
mid05tor12cn11: DEBUG State RequestedBMCTransition=xyz.openbmc_project.State.BMC.Transition.None
mid05tor12cn11: DEBUG State CurrentPowerState=xyz.openbmc_project.State.Chassis.PowerState.On
mid05tor12cn11: DEBUG State RequestedPowerTransition=xyz.openbmc_project.State.Chassis.Transition.Off
mid05tor12cn11: DEBUG State CurrentHostState=xyz.openbmc_project.State.Host.HostState.Running
mid05tor12cn11: DEBUG State RequestedHostTransition=xyz.openbmc_project.State.Host.Transition.On
mid05tor12cn11: on
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn11 off
mid05tor12cn11: off
monitorctrl::setNodeStatusAttributes called
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn11 stat
mid05tor12cn11: DEBUG State CurrentBMCState=xyz.openbmc_project.State.BMC.BMCState.Ready
mid05tor12cn11: DEBUG State RequestedBMCTransition=xyz.openbmc_project.State.BMC.Transition.None
mid05tor12cn11: DEBUG State CurrentPowerState=xyz.openbmc_project.State.Chassis.PowerState.Off
mid05tor12cn11: DEBUG State RequestedPowerTransition=xyz.openbmc_project.State.Chassis.Transition.Off
mid05tor12cn11: DEBUG State CurrentHostState=xyz.openbmc_project.State.Host.HostState.Off
mid05tor12cn11: DEBUG State RequestedHostTransition=xyz.openbmc_project.State.Host.Transition.On
mid05tor12cn11: off
```

Run ``rpower on``
```
# XCATBYPASS=1 rpower mid05tor12cn11 stat
mid05tor12cn11: DEBUG State CurrentBMCState=xyz.openbmc_project.State.BMC.BMCState.Ready
mid05tor12cn11: DEBUG State RequestedBMCTransition=xyz.openbmc_project.State.BMC.Transition.None
mid05tor12cn11: DEBUG State CurrentPowerState=xyz.openbmc_project.State.Chassis.PowerState.Off
mid05tor12cn11: DEBUG State RequestedPowerTransition=xyz.openbmc_project.State.Chassis.Transition.Off
mid05tor12cn11: DEBUG State CurrentHostState=xyz.openbmc_project.State.Host.HostState.Off
mid05tor12cn11: DEBUG State RequestedHostTransition=xyz.openbmc_project.State.Host.Transition.On
mid05tor12cn11: off
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn11 on
mid05tor12cn11: on
monitorctrl::setNodeStatusAttributes called
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn11 stat
mid05tor12cn11: DEBUG State CurrentBMCState=xyz.openbmc_project.State.BMC.BMCState.Ready
mid05tor12cn11: DEBUG State RequestedBMCTransition=xyz.openbmc_project.State.BMC.Transition.None
mid05tor12cn11: DEBUG State CurrentPowerState=xyz.openbmc_project.State.Chassis.PowerState.Off
mid05tor12cn11: DEBUG State RequestedPowerTransition=xyz.openbmc_project.State.Chassis.Transition.Off
mid05tor12cn11: DEBUG State CurrentHostState=xyz.openbmc_project.State.Host.HostState.Off
mid05tor12cn11: DEBUG State RequestedHostTransition=xyz.openbmc_project.State.Host.Transition.On
mid05tor12cn11: powering-on
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn11 stat
mid05tor12cn11: DEBUG State CurrentBMCState=xyz.openbmc_project.State.BMC.BMCState.Ready
mid05tor12cn11: DEBUG State RequestedBMCTransition=xyz.openbmc_project.State.BMC.Transition.None
mid05tor12cn11: DEBUG State CurrentPowerState=xyz.openbmc_project.State.Chassis.PowerState.On
mid05tor12cn11: DEBUG State RequestedPowerTransition=xyz.openbmc_project.State.Chassis.Transition.Off
mid05tor12cn11: DEBUG State CurrentHostState=xyz.openbmc_project.State.Host.HostState.Running
mid05tor12cn11: DEBUG State RequestedHostTransition=xyz.openbmc_project.State.Host.Transition.On
mid05tor12cn11: on
```

### CODE REVIEW

**Comment by whowutwut** [comment]
@xuweibj  I took this code change from your sandbox and put it onto the `briggs01` machine: xuweibj-status

The OFF case looks OK: 
```
[root@briggs01 vhu]# rpower mid05tor12cn05 state 
mid05tor12cn05: on
[root@briggs01 vhu]# rpower mid05tor12cn05 off 
mid05tor12cn05: off
[root@briggs01 vhu]# rpower mid05tor12cn05 state 
mid05tor12cn05: powering-off
[root@briggs01 vhu]# rpower mid05tor12cn05 state 
mid05tor12cn05: powering-off
[root@briggs01 vhu]# rpower mid05tor12cn05 state 
mid05tor12cn05: off
```

The ON case seems to not be working, as it's not reporting `powering-on`
```
[root@briggs01 vhu]# rpower mid05tor12cn05 on 
mid05tor12cn05: on
[root@briggs01 vhu]# rpower mid05tor12cn05 state 
mid05tor12cn05: off
[root@briggs01 vhu]# rpower mid05tor12cn05 state 
mid05tor12cn05: off
[root@briggs01 vhu]# rpower mid05tor12cn05 state 
mid05tor12cn05: off
[root@briggs01 vhu]# rpower mid05tor12cn05 state 
mid05tor12cn05: on
```

So something is not quite right...

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 1]
```diff
@@ -1398,33 +1411,39 @@ sub rpower_response {
                 # Chassis state is Off, but check if we can detect transition states
                 if ((defined($::OPENBMC_PWR) and ($::OPENBMC_PWR eq "YES")) and
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: I think if you implement this, we should remove the OPENBMC_PWR export variable....  and attempt to provide transition states all the time, for `off` and for `softoff` scenarios...

**Comment by xuweibj** [comment]
Hi @whowutwut , I have modified, please help to review again.
I copied code to briggs01 as /opt/xcat/lib/perl/xCAT_plugin/openbmc.pm.xw, you can use it to test.

This is my test result
```
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn05 stat
mid05tor12cn05: DEBUG State CurrentBMCState=xyz.openbmc_project.State.BMC.BMCState.Ready
mid05tor12cn05: DEBUG State RequestedBMCTransition=xyz.openbmc_project.State.BMC.Transition.None
mid05tor12cn05: DEBUG State CurrentPowerState=xyz.openbmc_project.State.Chassis.PowerState.On
mid05tor12cn05: DEBUG State RequestedPowerTransition=xyz.openbmc_project.State.Chassis.Transition.Off
mid05tor12cn05: DEBUG State CurrentHostState=xyz.openbmc_project.State.Host.HostState.Running
mid05tor12cn05: DEBUG State RequestedHostTransition=xyz.openbmc_project.State.Host.Transition.On
mid05tor12cn05: on
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn05 off
mid05tor12cn05: off
monitorctrl::setNodeStatusAttributes called
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn05 stat
mid05tor12cn05: DEBUG State CurrentBMCState=xyz.openbmc_project.State.BMC.BMCState.Ready
mid05tor12cn05: DEBUG State RequestedBMCTransition=xyz.openbmc_project.State.BMC.Transition.None
mid05tor12cn05: DEBUG State CurrentPowerState=xyz.openbmc_project.State.Chassis.PowerState.On
mid05tor12cn05: DEBUG State RequestedPowerTransition=xyz.openbmc_project.State.Chassis.Transition.Off
mid05tor12cn05: DEBUG State CurrentHostState=xyz.openbmc_project.State.Host.HostState.Running
mid05tor12cn05: DEBUG State RequestedHostTransition=xyz.openbmc_project.State.Host.Transition.On
mid05tor12cn05: powering-off
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn05 stat
mid05tor12cn05: DEBUG State CurrentBMCState=xyz.openbmc_project.State.BMC.BMCState.Ready
mid05tor12cn05: DEBUG State RequestedBMCTransition=xyz.openbmc_project.State.BMC.Transition.None
mid05tor12cn05: DEBUG State CurrentPowerState=xyz.openbmc_project.State.Chassis.PowerState.Off
mid05tor12cn05: DEBUG State RequestedPowerTransition=xyz.openbmc_project.State.Chassis.Transition.Off
mid05tor12cn05: DEBUG State CurrentHostState=xyz.openbmc_project.State.Host.HostState.Off
mid05tor12cn05: DEBUG State RequestedHostTransition=xyz.openbmc_project.State.Host.Transition.On
mid05tor12cn05: off
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn05 stat
mid05tor12cn05: DEBUG State CurrentBMCState=xyz.openbmc_project.State.BMC.BMCState.Ready
mid05tor12cn05: DEBUG State RequestedBMCTransition=xyz.openbmc_project.State.BMC.Transition.None
mid05tor12cn05: DEBUG State CurrentPowerState=xyz.openbmc_project.State.Chassis.PowerState.Off
mid05tor12cn05: DEBUG State RequestedPowerTransition=xyz.openbmc_project.State.Chassis.Transition.Off
mid05tor12cn05: DEBUG State CurrentHostState=xyz.openbmc_project.State.Host.HostState.Off
mid05tor12cn05: DEBUG State RequestedHostTransition=xyz.openbmc_project.State.Host.Transition.On
mid05tor12cn05: off
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn05 on
mid05tor12cn05: on
monitorctrl::setNodeStatusAttributes called
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn05 stat
mid05tor12cn05: DEBUG State CurrentBMCState=xyz.openbmc_project.State.BMC.BMCState.Ready
mid05tor12cn05: DEBUG State RequestedBMCTransition=xyz.openbmc_project.State.BMC.Transition.None
mid05tor12cn05: DEBUG State CurrentPowerState=xyz.openbmc_project.State.Chassis.PowerState.Off
mid05tor12cn05: DEBUG State RequestedPowerTransition=xyz.openbmc_project.State.Chassis.Transition.Off
mid05tor12cn05: DEBUG State CurrentHostState=xyz.openbmc_project.State.Host.HostState.Off
mid05tor12cn05: DEBUG State RequestedHostTransition=xyz.openbmc_project.State.Host.Transition.On
mid05tor12cn05: powering-on
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn05 stat
mid05tor12cn05: DEBUG State CurrentBMCState=xyz.openbmc_project.State.BMC.BMCState.Ready
mid05tor12cn05: DEBUG State RequestedBMCTransition=xyz.openbmc_project.State.BMC.Transition.None
mid05tor12cn05: DEBUG State CurrentPowerState=xyz.openbmc_project.State.Chassis.PowerState.On
mid05tor12cn05: DEBUG State RequestedPowerTransition=xyz.openbmc_project.State.Chassis.Transition.Off
mid05tor12cn05: DEBUG State CurrentHostState=xyz.openbmc_project.State.Host.HostState.Off
mid05tor12cn05: DEBUG State RequestedHostTransition=xyz.openbmc_project.State.Host.Transition.On
mid05tor12cn05: on (Chassis)
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn05 stat
mid05tor12cn05: DEBUG State CurrentBMCState=xyz.openbmc_project.State.BMC.BMCState.Ready
mid05tor12cn05: DEBUG State RequestedBMCTransition=xyz.openbmc_project.State.BMC.Transition.None
mid05tor12cn05: DEBUG State CurrentPowerState=xyz.openbmc_project.State.Chassis.PowerState.On
mid05tor12cn05: DEBUG State RequestedPowerTransition=xyz.openbmc_project.State.Chassis.Transition.Off
mid05tor12cn05: DEBUG State CurrentHostState=xyz.openbmc_project.State.Host.HostState.Running
mid05tor12cn05: DEBUG State RequestedHostTransition=xyz.openbmc_project.State.Host.Transition.On
mid05tor12cn05: on
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn05 stat
mid05tor12cn05: DEBUG State CurrentBMCState=xyz.openbmc_project.State.BMC.BMCState.Ready
mid05tor12cn05: DEBUG State RequestedBMCTransition=xyz.openbmc_project.State.BMC.Transition.None
mid05tor12cn05: DEBUG State CurrentPowerState=xyz.openbmc_project.State.Chassis.PowerState.On
mid05tor12cn05: DEBUG State RequestedPowerTransition=xyz.openbmc_project.State.Chassis.Transition.Off
mid05tor12cn05: DEBUG State CurrentHostState=xyz.openbmc_project.State.Host.HostState.Running
mid05tor12cn05: DEBUG State RequestedHostTransition=xyz.openbmc_project.State.Host.Transition.On
mid05tor12cn05: on
```

**Comment by whowutwut** [comment]
@xuweibj This still does not work as expected, maybe it's not accurate enough to implement..

I put your file onto `briggs01` and was going to run testing with it today....

```
[root@briggs01 ~]# rpower p9up state 
mid05tor12cn13: on
mid05tor12cn16: on
mid05tor12cn02: powering-on
mid05tor12cn11: on
mid05tor12cn15: on
mid05tor12cn05: on
mid05tor12cn17: on
mid05tor12cn18: on
```

So I didnt do anything for cn02 today, so it was odd it was powering on... so checking state: 

```
[root@briggs01 ~]# ssh root@172.11.139.2
root@172.11.139.2's password: 
root@witherspoon:~# obmcutil state
CurrentBMCState:    xyz.openbmc_project.State.BMC.BMCState.Ready
CurrentPowerState:  xyz.openbmc_project.State.Chassis.PowerState.Off
CurrentHostState:   xyz.openbmc_project.State.Host.HostState.Off
```

So this is the problem, when we hard off, the transition states are NOT updated correctly, so really it's up to xCAT states....  maybe we can't use status... it might not be reliable.  

```
[root@briggs01 ~]# lsdef mid05tor12cn02 -i status
Object name: mid05tor12cn02
    status=booted
[root@briggs01 ~]# XCATBYPASS=1 rpower mid05tor12cn02 state
mid05tor12cn02: DEBUG State CurrentBMCState=xyz.openbmc_project.State.BMC.BMCState.Ready
mid05tor12cn02: DEBUG State RequestedBMCTransition=xyz.openbmc_project.State.BMC.Transition.None
mid05tor12cn02: DEBUG State CurrentPowerState=xyz.openbmc_project.State.Chassis.PowerState.Off
mid05tor12cn02: DEBUG State RequestedPowerTransition=xyz.openbmc_project.State.Chassis.Transition.Off
mid05tor12cn02: DEBUG State CurrentHostState=xyz.openbmc_project.State.Host.HostState.Off
mid05tor12cn02: DEBUG State RequestedHostTransition=xyz.openbmc_project.State.Host.Transition.On
mid05tor12cn02: powering-on
```

So from this, I think we can't merge this PULL request.

**Comment by whowutwut** [comment]
I think I know what happened, this node I had to power pull the power,  so once the BMC booted, it is in OFF state.  And so since the transition states are not updated, it looks like it's "powering-on". 

So this this inconsistency, I think we should not push this in right now...

**Comment by xuweibj** [comment]
@whowutwut I think you are right. For this I think we could not know the correct status if the transition states are not updated. Will the firmware team support to change the transition states?

**Comment by whowutwut** [comment]
@xuweibj So far no luck.. will ask again tomorrow.

**Pull request closed without merging**